### PR TITLE
[Gen3] Fix Rock Smash in Safari Zone for RSE (partial #65)

### PR DESCRIPTION
**Pull request by Real96**
Fix for Rock Smash generation inside Safari Zone for RSE (partial #65)

### CODE REVIEW

File: Source/Core/Gen3/Searchers/WildSearcher3.hpp [hunk 1]
```diff
@@ -31,7 +31,7 @@ class WildSearcher3 : public WildSearcher
 {
 public:
     WildSearcher3() = default;
-    WildSearcher3(u16 tid, u16 sid, u8 genderRatio, Method method, const StateFilter &filter);
+    WildSearcher3(u16 tid, u16 sid, u8 genderRatio, Method method, const StateFilter &filter, bool isRSEVersion);
```
Comment (Admiral-Fish) [inline Source/Core/Gen3/Searchers/WildSearcher3.hpp:34]: Would prefer name of just `rse`

File: Source/Core/Gen3/Generators/WildGenerator3.hpp [hunk 2]
```diff
@@ -29,8 +29,11 @@ class WildGenerator3 : public WildGenerator
 {
 public:
     WildGenerator3() = default;
-    WildGenerator3(u32 initialAdvances, u32 maxAdvances, u16 tid, u16 sid, u8 genderRatio, Method method, const StateFilter &filter);
+    WildGenerator3(u32 initialAdvances, u32 maxAdvances, u16 tid, u16 sid, u8 genderRatio, Method method, const StateFilter &filter, bool isRSEVersion);
     std::vector<WildState> generate(u32 seed, const EncounterArea3 &encounterArea) const;
+
+private:
+    bool isRSEVersion;
```
Comment (Admiral-Fish) [inline Source/Core/Gen3/Generators/WildGenerator3.hpp:36]: Would prefer name of just `rse`

File: Source/Core/Gen3/Searchers/WildSearcher3.hpp [hunk 3]
```diff
@@ -41,6 +41,7 @@ class WildSearcher3 : public WildSearcher
 private:
     RNGCache cache;
     EncounterArea3 encounterArea;
+    bool isRSEVersion;
```
Comment (Admiral-Fish) [inline Source/Core/Gen3/Searchers/WildSearcher3.hpp:44]: Would prefer name of just `rse`

File: Source/Core/Gen3/Searchers/WildSearcher3.cpp [hunk 4]
```diff
@@ -24,9 +24,10 @@
 #include <Core/RNG/LCRNG.hpp>
 #include <Core/Util/EncounterSlot.hpp>
 
-WildSearcher3::WildSearcher3(u16 tid, u16 sid, u8 genderRatio, Method method, const StateFilter &filter) :
+WildSearcher3::WildSearcher3(u16 tid, u16 sid, u8 genderRatio, Method method, const StateFilter &filter, bool isRSEVersion) :
     WildSearcher(tid, sid, genderRatio, method, filter), cache(method), searching(false), progress(0)
```
Comment (Admiral-Fish) [inline Source/Core/Gen3/Searchers/WildSearcher3.cpp:28]: Same comment as generator, move line outside of constructor to `rse(rse)`

File: Source/Core/Gen3/Searchers/WildSearcher3.cpp [hunk 5]
```diff
@@ -97,6 +98,21 @@ std::vector<WildState> WildSearcher3::search(u8 hp, u8 atk, u8 def, u8 spa, u8 s
         return states;
     }
 
+    bool isRSESafariLocation = encounterArea.isRSESafariZone() && isRSEVersion;
```
Comment (Admiral-Fish) [inline Source/Core/Gen3/Searchers/WildSearcher3.cpp:101]: `bool rseSafariRockSmash = encounterArea.isRseSafariZone() && rse && encounter == Encounter::RockSmash;`

Removes needs for the switch and stuff

File: Source/Core/Gen3/EncounterArea3.hpp [hunk 6]
```diff
@@ -30,6 +30,7 @@ class EncounterArea3 : public EncounterArea
     u8 calcLevel(u8 index, u16 prng) const;
     u8 calcLevel(u8 index) const;
     u8 getEncounterRate() const;
+    bool isRSESafariZone() const;
```
Comment (Admiral-Fish) [inline Source/Core/Gen3/EncounterArea3.hpp:33]: `bool rseSafariZone() const;`

**Review by Admiral-Fish — CHANGES_REQUESTED** [review]
General question, I assume you verified both generator/searcher are accurate?

**Comment by Real96** [comment]
> General question, I assume you verified both generator/searcher are accurate?

Yep, another user tested it too!

File: Source/Core/Gen3/Searchers/WildSearcher3.cpp (hunk 5 above)
Comment (Real96) [reply Source/Core/Gen3/Searchers/WildSearcher3.cpp:101]: fishing will need this too, can i leave this? or do i add another bool for fishing?

File: Source/Core/Gen3/Generators/WildGenerator3.hpp [hunk 7]
```diff
@@ -29,8 +29,11 @@ class WildGenerator3 : public WildGenerator
 {
 public:
     WildGenerator3() = default;
-    WildGenerator3(u32 initialAdvances, u32 maxAdvances, u16 tid, u16 sid, u8 genderRatio, Method method, const StateFilter &filter);
+    WildGenerator3(u32 initialAdvances, u32 maxAdvances, u16 tid, u16 sid, u8 genderRatio, Method method, const StateFilter &filter, bool isRSEVersion);
```
[the remark this answers is not in the document]
Comment (Real96) [reply Source/Core/Gen3/Generators/WildGenerator3.hpp]: done

File: Source/Core/Gen3/Generators/WildGenerator3.cpp [hunk 8]
```diff
@@ -29,9 +29,10 @@
 #include <functional>
 
 WildGenerator3::WildGenerator3(u32 initialAdvances, u32 maxAdvances, u16 tid, u16 sid, u8 genderRatio, Method method,
-                               const StateFilter &filter) :
+                               const StateFilter &filter, bool isRSEVersion) :
     WildGenerator(initialAdvances, maxAdvances, tid, sid, genderRatio, method, filter)
```
[the remark this answers is not in the document]
Comment (Real96) [reply Source/Core/Gen3/Generators/WildGenerator3.cpp]: done

File: Source/Core/Gen3/Generators/WildGenerator3.cpp [hunk 9]
```diff
@@ -29,9 +29,10 @@
...
 {
+    this-> isRSEVersion = isRSEVersion;
```
[the remark this answers is not in the document]
Comment (Real96) [reply Source/Core/Gen3/Generators/WildGenerator3.cpp]: done

File: Source/Core/Gen3/Generators/WildGenerator3.cpp [hunk 10]
```diff
@@ -42,6 +43,8 @@ std::vector<WildState> WildGenerator3::generate(u32 seed, const EncounterArea3 &
     rng.advance(initialAdvances + offset);
 
     u16 rate = encounterArea.getEncounterRate() * 16;
+    bool isRSESafariLocation = encounterArea.isRSESafariZone() && isRSEVersion; // RockSmash encounters have different rng calls inside RSE Safari Zone,
```
[the remark this answers is not in the document]
Comment (Real96) [reply Source/Core/Gen3/Generators/WildGenerator3.cpp]: done

File: Source/Core/Gen3/Searchers/WildSearcher3.hpp (hunk 1 above)
Comment (Real96) [reply Source/Core/Gen3/Searchers/WildSearcher3.hpp]: done

File: Source/Core/Gen3/Generators/WildGenerator3.hpp (hunk 2 above)
Comment (Real96) [reply Source/Core/Gen3/Generators/WildGenerator3.hpp]: done

File: Source/Core/Gen3/Searchers/WildSearcher3.hpp (hunk 3 above)
Comment (Real96) [reply Source/Core/Gen3/Searchers/WildSearcher3.hpp]: done

File: Source/Core/Gen3/Searchers/WildSearcher3.cpp (hunk 4 above)
Comment (Real96) [reply Source/Core/Gen3/Searchers/WildSearcher3.cpp]: done

File: Source/Core/Gen3/EncounterArea3.hpp (hunk 6 above)
Comment (Real96) [reply Source/Core/Gen3/EncounterArea3.hpp]: done

File: Source/Core/Gen3/Searchers/WildSearcher3.cpp (hunk 5 above)
Comment (Real96) [reply Source/Core/Gen3/Searchers/WildSearcher3.cpp]: > Do they do similar things to the RNG?

yes, basically the same thing

Comment (Real96) [reply Source/Core/Gen3/Searchers/WildSearcher3.cpp]: I changed it as requested, we'll discuss this in later fishing pr!

**Pull request merged**